### ISSUE TITLE
Idea: Sagas 'returning' promises from middleware (to support Server Side Rendering)

### ISSUE THREAD

**Issue by quicksnap** [yelouafi/redux-saga#149]
Last night I was working with sagas, doing standard-fare authentication flow. However, I use server rendering, and sagas seem to be lacking for this.

In my existing apps, I do something like `dispatch(login(email, password)).then(/*...*/)`. I then proceed to render the route and send a response. Of course, with sagas it's not as simple. 

I have read #13 and `runSagas` is a good option to have. However, what if we consider an effect that alters the response of `dispatch` on certain actions?

Let's call it `takeDefer` for now. A better name would be nice if this idea is viable

```js
function doLogin(email, password) {
  const promise = dispatch({ type: 'LOGIN',  payload: { email, password } });
  promise.then(/*...*/).catch(/*...*/);
}

function * login() {
  while (true) {
    // Right now, dispatch(LOGIN) would return a POJO
    // Once the generator yields `takeDefer` to the middleware, it latches
    // on and alters the return value of `dispatch(LOGIN)`
    
    const task = yield takeDefer('LOGIN'); 

    // At this point, subsequent dispatch(LOGIN) would not return a promise
    
    const { username, password } = task.action.payload;
    const authTask = yield fork(authorize, username, password);

    const { fail } = yield race({
      success: take('LOGIN_SUCCESS'),
      fail: take('LOGIN_FAIL'),
    })

    if (fail) {
      task.reject('Login has failed')
    } else {
      task.resolve('Login complete!')
    }

    yield take('LOGOUT');
  }
}
```

If there were multiple `takeDefer('LOGIN')` effects concurrently waiting, the promise returned from the middleware would be `Promise.all(...)`. The array of promises could be made available as well.

I believe this is possible, and would love to give a whack at a PR. However, does this sound sane? Does this kind of usage of sagas break their intent?

From a practical standpoint, being able to leverage promises like this could be very useful for server rendering--I could reuse the same client-saga on the server, and jump out of it once I have what is needed. Also, it would help with CPS methods such as `onEnter` in React Router.

**Comment by j** [comment]
I'm probably being an annoying chime-in, but I just refactored from redux-thunk and it wasn't too painful... until I got to server-rendering for redux-saga.  I modified to fit #13 and and for me, the promise is completing before my api responds.  So far I thought I basically copy/pasted, but I guess I can dig further later.  None-the-less, it was slightly painful to do.

**Comment by pavelkornev** [comment]
What we need is an indicator when sagas are not running. Instead of returning Promises from sagas itself i propose ```sagaMiddleware.done``` similarly to ```done``` property of the ```tasks```.

Then we can use client and server side sagas on both sides and isolate our components from sagas existence (see my solution in #13).

The only problem i see so far is endless cycles in sagas which we use to demonization:
```
while(true) { ... }
```
As a work around ```sagaMiddleware``` can track ```take``` effects internally and and resolve global promise when all registered sagas in the first stage; this means all sagas are waiting for an initial ```take``` effect.

**Comment by yelouafi** [comment]
how do we actually handle universal rendering using `redux-thunk` and `react-router` ?  I tried to look at the redux universal demo project but didn't find a react-router example (which uses match). It only uses a generic handler which [always fires the same api call](https://github.com/reactjs/redux/blob/master/examples/universal/server/server.js#L33). What I expected is an of example of a componentWillMount -> dispatch data-load tasks -> wait completion -> render process.

@pavelkornev looked at your universal example. In [waitAll](https://github.com/pavelkornev/redux-saga/blob/master/examples/real-world-universal/sagas/waitAll.js) you fork multiple tasks then immediately join them. You can achieve this directly with parallel calls.

```javascript
export default (sagas) => function* () {
  yield sagas.map(([saga, ...params]) => call(saga, ...params))
}
```

**Comment by yelouafi** [comment]
I found this one
https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/containers/App/App.js

It uses a kind of async Connect that collect load-time promises on the server from mounted components. It provides a `loadOnServer` helper which apparently waits for all collected promises before resolving.

Since it uses redux-thunk. Async-Connected components can simply dispatch thunks and pass the returned promises to asyncConnect.

It seems that@pavelkornev solution is similar. Components define load-time tasks which get collected on the match method. But seems this is not quite the Saga way of doing things

The watcher model used actually doesnt play nice with the server processing model. 

The while(true) loop isnt the real issue. We can break it with some server specific event. Or we can place it in a conditional isServer test.

The issue is that the middleware starts all sagas while in the server we need only to start the sagas that handles the matching route's actions. middleware.done will wait for all started sagas but since only current route's actions will fire from mounted components, sagas waiting for actions on unmounted components will stay blocked on their `take` effect. So the middleware.done will never resolve

**Comment by pavelkornev** [comment]
@yelouafi

1. Yes, [react-redux-universal-hot-example](https://github.com/erikras/react-redux-universal-hot-example) uses [redux-async-connect](https://github.com/Rezonans/redux-async-connect). But it actually doesn't collect promises "on the server from **mounted** components". It collects promises from statics of the resolved with ```react-router``` components. Where is difference? If we collect from mounted components that means we called ```renderToString```, then wait while all collected promises resolved and then render again, so it's rendering twice. On the other hand, downside of collecting promises from resolved components via react-router is that we collect promises only from them and we don't have access for possible children which means we obligate parent components to know which data need their children.

2. About middleware we are talking about almost same problem but on different angle. Difference between two approaches is: you say &mdash; we should not register unnessary sagas on the server which give us a problem to know which ones we need and it's not that trivial as it seems, mine &mdash; start all sagas and kill those are not using.

Starting all sagas is a way easier then determine which ones we need for our rendering tree. That's why we do so on the client and it works like a charm. So why it could not be same useful on the server?

But there is one little thing. How to determine which ones we need to kill out of our waiting list? As it proposed above we can kill those are at the initial ```take``` effect. So, on the server side components will emit events to run necessary sagas, then we filter list of registered sagas by killing Idle sagas and return promise which resolves when working sagas finish their work.

**Comment by quicksnap** [comment]
Also, with my idea of `takeDefer` enabling saga middleware to return promises, we don't even need to wait for sagas to complete. We may simply render once we have received all deferred requests necessary.

**Comment by yelouafi** [comment]
@quicksnap I think the takeDefer solution a bit complicated. I prefer also to not modify the core unless it's necessary.

That been said. It seems the 2 solutions are centered around the same issue (@pavelkornev correct me if I'm mistaken) which is tracking the resolution of the first take effect. Either to turn it into a promise in the 1st solution or to filter out and kill sagas who haven't resolved their take effect in the current event loop.

If the above is correct there is an alternative option which doesn't involve changing the core and also more flexible which is saga monitoring Api. 

All sagas fire actually events on each effect creation/resolution.

@quicksnap We can create a server saga monitor which will track take effects fired within the current event loop and their resolution.  Since a saga monitor is a middleware it can transform the first takes into promisea

@pavelkornev We can also create a saga monitor which directly tracks all tasks forked by watchers in the current event loop. So we dont need to filter out and cancel. In the saga monitor we can simply collect all forked tasks and wait for all of them.

**Comment by quicksnap** [comment]
@yelouafi Interesting! I didn't see the sagaMonitor example before. I will look into it and see if it can achieve what's needed. It sounds promising..

**Comment by yelouafi** [comment]
Another angle of view is: on the server we want to run some tasks before rendering but we do not know those tasks because they are encapsulated within watchers.

@pavelkornev solution solves this but couples the components with sagas.

So a possible solution is:

- Use component statics to define only load-time actions

- write a custom saga driver: it will iterate on all watchers and get their first take effects.

- filters to only keep watchers whose take pattern matches the load-time actions

- resume each watcher with its matching action

- get the next fork effect which gives us infos about the task to run

- run all collected tasks and wait for them

This has the advantage of coupling components only with actions. And we do not have to call renderToString twice. But as @pavelkornev mentioned it works only if your Router mount points matches your Connect mount points

-

**Comment by pavelkornev** [comment]
@yelouafi 'saga driver' sounds promising. Could you help with implementation?

**Comment by yelouafi** [comment]
Note that the driver expect the watchers to yield in a specific order; also can extend to support other action patterns (array, functions)

```javascript

import { take, fork } from 'redux-saga/effects'
import{ asEffect } from 'redux-saga/utils'

function* commonTask() { }

function* watchCommonTask() {
  while(true) {
    const action = yield take('COMMON_TASK')
    yield fork(commonTask, action.data)
  }
}

function* loadUser(action) {}

function* watchLoadUser() {
  while(true) {
    const action = yield take('LOAD_USER')
    yield fork(loadUser, action.user)
  }
}

function* loadRepo(action) {}

function* watchLoadRepo() {
  while(true) {
    const action = yield take('LOAD_REPO')
    yield fork(loadRepo, action.repo)
  }
}


function watcherDriver(watcher, filterActions) {

  const gen = watcher()

  let { value: effect } = gen.next()
  const pattern = asEffect.take(effect)
  if(pattern !== undefined) {
    for (var i = 0; i < filterActions.length; i++) {
      const action = filterActions[i]
      if(pattern === action.type) {
        let { value: effect } = gen.next(action)
        const taskDef = asEffect.fork(effect)
        if(taskDef !== undefined) {
          return {
            action,
            saga: taskDef.fn,
            args: taskDef.args
          }
        }
      }
    }
  }

}

function getLoadTasks(watchers, filterActions) {
  return watchers.map(watcher => watcherDriver(watcher, filterActions))
                 .filter(taskDef => taskDef !== undefined)
}


console.log(getLoadTasks(
  [watchLoadUser, watchLoadRepo, watchCommonTask],
  [{type: 'LOAD_REPO', user: 'user-1'}, {type: 'COMMON_TASK', data: 'data-1'}]
))
```

**Comment by yelouafi** [comment]
@quicksnap @pavelkornev have something ?

**Comment by quicksnap** [comment]
@yelouafi I've been testing out Sagas on a side project. To have server rendering work without much pain is important, as I want to make traction on the project in my spare time. Right now I've been inclined to fall back on my more familiar pattern of promise middleware combined with thunks.

Having to work in a solution outside of core is a drawback for me, as it increases the complexity of my project, and I am not certain yet that it will yield the benefit of lowering the overall complexity of action orchestration. It feels like it could cause more code or confusion around what's happening in either the server or client environments.

Honestly, these thoughts are coming from my perspective of not putting in enough time/thought into this problem space. I really appreciate your examples in this thread, and I think they could work to help the problem. However, have an official way of supporting server rendering that doesn't involve separate client/server sagas is crucial to me.

Should we create a new issue to think about best solutions to this? I _really_ want to use sagas, and if we can work a solution that either gets into core or gracefully solves the problem, I'd be happy to put time into helping!

Thanks for your work so far on all this, it's great!

**Comment by pavelkornev** [comment]
@quicksnap @yelouafi
I'm currently working on that kind of solution which makes it possible to run same sagas on both sides without pain. It's based on idea of orchestration sagas in custom Saga Monitor. Hopefully i can show it soon.

**Comment by quicksnap** [comment]
@pavelkornev Thanks for taking it up! I wish I could devote more time to it right now..

**Comment by yelouafi** [comment]
@quicksnap @pavelkornev I've a proposed solution here. In https://github.com/yelouafi/redux-saga/issues/78#issuecomment-203886961 I propose that a parent is terminated only when all attached forked tasks terminate. If we combine this with a special `END` action that I intend to introduce. We can have soemthing like this

```javascript
function* watcher() {
  let action
  while((action = take('ACTION') === END) {
    yield fork(worker, action.payload)
  }
}

function* rootSaga() {
  yield [
    fork(watcher),
    fork(anotherWatcher)
  ]
}


// on the server rendering code, inside match

const sagaRunner = createSagaMiddleware() // no saga is passed here
const store = createStore(reducer, applyMiddleware(sagaRunner))
const rootTask = sagaRunner.run(rooSaga)

renderToString(RootComponent)
// will break the while loop on watchers
store.dispatch(END)

// rootTask will resolve only when all tasks in the program are terminated
rootTask.done.then(/* render and send state + html to the client */)
```

`END` is a special action, it notifies all takers that no more actions will be dispatched. And since a saga won't terminate until all forked tasks terminate, we can use the `done` promise of the root saga  to determine when all forked tasks are completed. Similarly the root saga will abort on the first error encountered by a forked task so we can use this to send an error to the client.

W'll be rendering twice of course on the server, but The main advantage here is that the same saga code would run on the server and client

**Comment by quicksnap** [comment]
@yelouafi at first blush, I like this approach. I'll try and provide some more thoughtful feedback soon. 

Thanks for putting so much effort into this! If you happen to be going to React Europe, I'll buy you a beer!

**Comment by yelouafi** [comment]
see #255
